changelogs/unreleased/gh-10180-upgrade-from-1-6.md

```diff
@@ -0,0 +1,4 @@
+## bugfix/upgrade
```

Comment:
> 2. Start to automatically change format of user spaces: previously it
   was {'a', type = 'number'}, now it's {name = 'a', type = 'number'}.
   Make sure, that space format is encoded as array.

Are you sure it wasn't possible to set format like `{{'a', 'number'}}`? I see this syntax is supported on new tarantool versions, but the format is converted to `{{name='a', type='number'}}` internally.

Also I don't understand this sentence: `Make sure, that space format is encoded as array.`

Reply:
> Are you sure it wasn't possible to set format like {{'a', 'number'}}? I see this syntax is supported on new tarantool versions, but the format is converted to {{name='a', type='number'}} internally.

Back in 1.6 it was also possible to use both variants. The problem there, is that it have not been converted:

```lua
src/tarantool: version 1.6.9-94-ga5da60bdd
type 'help' for interactive help
<...>
tarantool> box.schema.space.create('test', {format = {{name = 'a', type = 'number'}}})
<...>
arantool> box.schema.space.create('test2', {format = {{'a', type = 'number'}}})
<...>
tarantool> box.space._space:select()
<...>
  - [512, 1, 'test', 'memtx', 0, {}, [{'name': 'a', 'type': 'number'}]]
  - [513, 1, 'test2', 'memtx', 0, {}, [{1: 'a', 'type': 'number'}]]
```

Without changing the format of user spaces, we fail here:

https://github.com/tarantool/tarantool/blob/6ae595f5da606478223d69b641213444766bca85/src/box/field_def.c#L512-L516

```text
"object_id","type":"unsigned"},{"name":"privilege","type":"unsigned"}]
[001] server | 2024-10-31 13:18:51.814 [19517] main/104/server_instance.lua/box.upgrade I> Update space '_cluster' format: new format [{"name":"id","type":"unsigned"},{"name":"uuid","type":"string"}]
[001] server | 2024-10-31 13:18:51.814 [19517] main/104/server_instance.lua memtx_engine.cc:450 E> error at request: {type: 'INSERT', replica_id: 0, lsn: 33, space_id: 280, index_id: 0, tuple: [514, 1, "format", "memtx", 0, {}, [{1: "a", "type": "number"}, {1: "b", "type": "num"}, {1: "c", "type": "array"}, {1: "d", "type": "array"}]]}
[001] server | 2024-10-31 13:18:51.814 [19517] main/104/server_instance.lua field_def.c:514 E> ER_WRONG_SPACE_FORMAT: Wrong space format field 1: expected a map with string keys
[001] server | 2024-10-31 13:18:51.814 [19517] main/104/server_instance.lua F> can't initialize storage: Wrong space format field 1: expected a map with string keys
[001] server | 2024-10-31 13:18:51.814 [19517] main/104/server_instance.lua F> can't initialize storage: Wrong space format field 1: expected a map with string keys
[001] not ok 1	box-luatest.gh_10180_upgrade_from_1_6_9.test_upgrade
[001] #   ...ng/tnt/tarantool/test-run/lib/luatest/luatest/server.lua:539: Process is terminated when waiting for "server is ready" condition for server (alias: server, workdir: server-8qNUtZhveKJA, pid: 19517)
[001] #   stack traceback:
[001] #   	...ol/test/box-luatest/gh_10180_upgrade_from_1_6_9_test.lua:10: in function <...ol/test/box-luatest/gh_10180_upgrade_from_1_6_9_test.lua:6>
[001] #   	...
[001] #   	[C]: in function 'xpcall'
[001] # Ran 1 tests in 0.334 seconds, 0 succeeded, 1 errored
[001] 
```

Reply:
Thanks for addressing this issue!